### DISCUSSION TITLE
Does sycl have a feature like cudaStreamBeginCapture to create graph dependency automatically when using taskflow?

### DISCUSSION THREAD

**Discussion by mazhaojia123** [intel/llvm#12950] [Q&A] [answered]
Hi, 

Does sycl have a feature like cudaStreamBeginCapture to create graph dependency automatically when using taskflow?

Is there any code sample like following code using sycl with taskflow?

```cpp
# A simple cuda code sample to create graph.
cudaStreamBeginCapture(default_stream, cudaStreamCaptureModeGlobal);
kernel();
cudaStreamEndCapture(default_stream, &graph);
cudaGraphInstantiate(&instance, graph, NULL, NULL, 0);
```

BTW, I want to offload my SYCL kernels to nvidia GPUs. 
So it seems that I can only use the taskflow instead of new SYCL graph feature.

Thank you.

**Comment by bader** [comment]
Tagging @intel/sycl-graphs-reviewers team to answer.

**Reply by EwanC** [reply, marked as answer]
Hello, thanks for the question. The [sycl_ext_oneapi_graph](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc) extension to DPC++ should provide an equivalent feature using code like:

```cpp
namespace sycl_ext = sycl::ext::oneapi::experimental;

sycl::queue q;
sycl_ext::command_graph graph(q);

graph.begin_recording(q); // equivalent to cudaStreamBeginCapture

// queue submissions recorded here, equivalent to `kernel()`
// e.g. q.submit(/*kernel*/);

graph.end_recording(); // equivalent to cudaStreamEndCapture

auto exec_graph = graph.finalize(); // equivalent to cudaGraphInstantiate
```


sycl_ext_oneapi_graph supports the CUDA SYCL backend by using CUDA-Graph in the implementation, so you will be able to use an NVIDIA GPU with the extension. I can't comment on taskflow specifics, but please let us know of any properties of that application you think are relevant to compatibility with sycl_ext_oneapi_graph

**Reply by mazhaojia123** [reply]
Thank you, @EwanC. It's very helpful.
